File: cpp/src/discovery.cpp
// Auto-detect paired BMAP devices via bluetoothctl (Linux).
#include "discovery.h"

#include <array>
#include <cstdio>
#include <memory>
#include <regex>
#include <sstream>
#include <string>
#include <vector>

namespace bmap {

static const char* BMAP_UUID = "00000000-deca-fade-deca-deafdecacaff";

static std::string exec(const std::string& cmd) {
    std::array<char, 256> buf;
    std::string result;
    std::unique_ptr<FILE, int(*)(FILE*)> pipe(popen(cmd.c_str(), "r"), pclose);
    if (!pipe) return "";
    while (fgets(buf.data(), buf.size(), pipe.get())) {
        result += buf.data();
    }
    return result;
}

// Known BMAP devices (from https://downloads.bose.com/lookup.xml):
//   0x4017 kleos    — QuietComfort 35           → qc35
//   0x4020 baywolf  — QuietComfort 35 II        → qc35
//   0x4024 goodyear — NC Headphones 700         (unsupported)
//   0x4060 olivia   — QC Earbuds II             (unsupported)
//   0x4061 vedder   — QuietComfort 45           (unsupported)
//   0x4063 edith    — Ultra Open Earbuds        (unsupported)
//   0x4075 prince   — QC Ultra Earbuds          (unsupported)
//   0x4082 wolverine — QC Ultra Headphones      → qc_ultra2
static std::string detect_device_type(const std::string& info) {
    std::regex modalias_re(R"(Modalias:\s*bluetooth:v[0-9A-Fa-f]{4}p([0-9A-Fa-f]{4}))");
    std::smatch match;
    if (std::regex_search(info, match, modalias_re)) {
        unsigned int product_id = std::stoul(match[1].str(), nullptr, 16);
        if (product_id == 0x4082) return "qc_ultra2";                  // wolverine
        if (product_id == 0x4020 || product_id == 0x4017) return "qc35"; // baywolf / kleos
    }
    return "qc_ultra2";
}

std::optional<std::pair<std::string, std::string>> find_bmap_device() {
    auto output = exec("bluetoothctl devices Paired 2>/dev/null");
    std::istringstream stream(output);
    std::string line;

    struct Candidate { std::string mac; std::string device_type; bool connected; };
    std::vector<Candidate> candidates;

    while (std::getline(stream, line)) {
        auto first_space = line.find(' ');
        if (first_space == std::string::npos) continue;
        auto second_space = line.find(' ', first_space + 1);
        if (second_space == std::string::npos) continue;
        auto mac = line.substr(first_space + 1, second_space - first_space - 1);

        auto info = exec("bluetoothctl info " + mac + " 2>/dev/null");

        bool is_audio = (info.find("audio-headset") != std::string::npos ||
                         info.find("audio-headphones") != std::string::npos);
        bool has_bmap = info.find(BMAP_UUID) != std::string::npos;
        if (!(is_audio && has_bmap)) continue;

        bool connected = info.find("Connected: yes") != std::string::npos;
        candidates.push_back({mac, detect_device_type(info), connected});
    }

    // Prefer connected devices
    for (auto& c : candidates) {
        if (c.connected) return std::make_pair(c.mac, c.device_type);
    }
    if (!candidates.empty()) {
        return std::make_pair(candidates[0].mac, candidates[0].device_type);
    }
    return std::nullopt;
}

} // namespace bmap

File: rust/src/discovery.rs
//! Auto-detect paired BMAP devices via bluetoothctl (Linux).

use std::process::Command;

/// BMAP service UUID found in SDP records.
pub const BMAP_UUID: &str = "00000000-deca-fade-deca-deafdecacaff";

/// A discovered BMAP device.
#[derive(Debug, Clone)]
pub struct DiscoveredDevice {
    pub mac: String,
    pub device_type: String,
    pub connected: bool,
}

/// Auto-detect a paired, connected BMAP-capable Bluetooth device.
///
/// Prioritizes connected devices. Returns (mac, device_type), or None.
pub fn find_bmap_device() -> Option<(String, String)> {
    let candidates = scan_paired_devices();

    // Prefer connected
    for d in &candidates {
        if d.connected {
            return Some((d.mac.clone(), d.device_type.clone()));
        }
    }
    // Fall back to first paired
    candidates.first().map(|d| (d.mac.clone(), d.device_type.clone()))
}

/// Scan all paired Bluetooth devices for BMAP-capable headphones.
pub fn scan_paired_devices() -> Vec<DiscoveredDevice> {
    let mut candidates = Vec::new();
    let output = match Command::new("bluetoothctl")
        .args(["devices", "Paired"])
        .output()
    {
        Ok(o) => String::from_utf8_lossy(&o.stdout).into_owned(),
        Err(_) => return candidates,
    };

    for line in output.lines() {
        let parts: Vec<&str> = line.splitn(3, ' ').collect();
        if parts.len() < 2 {
            continue;
        }
        let mac = parts[1];

        let info = match Command::new("bluetoothctl")
            .args(["info", mac])
            .output()
        {
            Ok(o) => String::from_utf8_lossy(&o.stdout).into_owned(),
            Err(_) => continue,
        };

        let is_audio = info.contains("audio-headset") || info.contains("audio-headphones");
        let has_bmap = info.contains(BMAP_UUID);
        if !(is_audio && has_bmap) {
            continue;
        }

        let connected = info.contains("Connected: yes");
        let device_type = detect_device_type(&info);

        candidates.push(DiscoveredDevice {
            mac: mac.to_string(),
            device_type,
            connected,
        });
    }
    candidates
}

/// Detect device type from Modalias product ID.
fn detect_device_type(info: &str) -> String {
    // Modalias format: bluetooth:vXXXXpYYYYdZZZZ
    for line in info.lines() {
        let trimmed = line.trim();
        if trimmed.starts_with("Modalias:") {
            if let Some(p_pos) = trimmed.find('p') {
                let id_str = &trimmed[p_pos + 1..p_pos + 5];
                if let Ok(product_id) = u16::from_str_radix(id_str, 16) {
                    return match product_id {
                        0x4082 => "qc_ultra2".to_string(),
                        0x4020 | 0x400C => "qc35".to_string(),
                        _ => "qc_ultra2".to_string(),
                    };
                }
            }
        }
    }
    "qc_ultra2".to_string()
}
